color and font support gtk (linux)

Vipul-Cariappa opened the pull request:
<!--- Describe your changes in detail -->
Added Support to change background color, text color and font size, family, style, weight, etc., in label and button widgets in Gtk backend.
<!--- What problem does this change solve? -->
<!--- If this PR relates to an issue, include Refs #XXX or Fixes #XXX -->
Fixes #1200 
Refs #949 
## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

Note: I have research on changing background color of a `Gtk.Fixed` widget and could not find a way to do this (But I have included the code which tries to change background color and it is not working). Looks like using some other placeholder widget is required if we desire to change background color of `Toga.Box` widget for linux backend.

Vipul-Cariappa commented:
The tests failed because I am using f-string which was added in python3.6. And the community is planning to remove python3.5 support. Refs #1199 #1201 
@freakboy3742 could you please advise if I need to change f-string to `str.format(...)`

Vipul-Cariappa commented:
### Wanted Advise
The code to change font in label and button widgets are the same. So instead of repeating the same code I was thinking of making a function in `toga_gtk/fonts.py` which takes in `StyleContext` and `value` and applies the value to the widget.
Similar function for changing colors in `toga_gtk/colors.py`.
If the above mentioned sounds good I will start working on it.

freakboy3742 commented:
@Vipul-Cariappa If there is repeated code, it should absolutely be pulled into a utility method, not copy-pasted everywhere. Since CSS-based styles are a feature of GTK itself, it would make sense to collect those in the libs folder of the GTK backend (possibly in a new "style" module).

As you've noted, *right now*, we support Python 3.5 - so we can't use f-strings; however, we're in the process of deprecating Python 3.5, so I'd put that at the end of your todo list. There's a possibility that by the time you get around to that problem, it may no longer be a problem :-)

I also note that you've added a block of code with an explicit "This doesn't work" comment. We're unlikely to merge code that doesn't work - you'll need to either remove that code, or work out the fixes required.

Lastly - you'll note that the test suite is currently failing the flake8 check. flake8 checks for consistent code style - things like indentation and spacing. If you run flake8 locally (or check the logs on CI), you can get a list of the fixes that are needed.

Vipul-Cariappa commented:
> If there is repeated code, it should absolutely be pulled into a utility method, not copy-pasted everywhere. Since CSS-based styles are a feature of GTK itself, it would make sense to collect those in the libs folder of the GTK backend (possibly in a new "style" module).

I have transferred most of the repeated code to `style.py` in `libs` folder

> I also note that you've added a block of code with an explicit "This doesn't work" comment. We're unlikely to merge code that doesn't work - you'll need to either remove that code, or work out the fixes required.

I have completely removed the code which does not work. (Replaced with `not_implemented`)
_I could not find a way to change background color of `Gtk.Fixed` widget. We may have to change `Gtk.Fixed` to `Gtk.Box` or some other widget if we desire to add the ability to change background color of `Toga.Box`._

The [winforms pythonnet install](https://github.com/beeware/toga/pull/1203/checks?check_run_id=1843975000#step:4:82) is failing which is causing the tests to fail. I have not changed anything from winforms. And it looks like pythonnet released new version [`v2.5.2`](https://github.com/pythonnet/pythonnet/releases/tag/v2.5.2) 9 hours ago from the time I am writing this.

@freakboy3742  If there is any other changes required please inform. I am happy to help.

Vipul-Cariappa commented:
![font_color_gtk](https://user-images.githubusercontent.com/55694469/107232277-3dfac900-6a47-11eb-9c8d-ea9281965ed1.png)
@freakboy3742 The above image shows the changes.

Vipul-Cariappa commented:
@freakboy3742 this branch is ready to be merged. If any changes is required I am happy to help.

freakboy3742 commented on src/gtk/toga_gtk/libs/style.py:
```diff
@@ -0,0 +1,49 @@
+        style_provider,
+        Gtk.STYLE_PROVIDER_PRIORITY_USER,
+    )
+    style_context.add_class("custom-style")
```
If I'm reading this right, GTK styling is based on the idea that every app has a CSS stylesheet (possibly building on a "system theme" stylesheet); and then individual widgets can tweak that style by attaching an encoded style that defines a class, and attaching that class to the widget.

If that's the case, I have 2 questions/clarifications:

1) Would it make sense to use a name like `toga` as the custom class, rather than the generic class name?I'm thinking mostly for debugging purposes, being able to identify where a style change comes from. `custom-style` is sufficiently generic that it might not be obvious where the style is coming from.

2) Is there any potential for namespace collision here? i.e., if I have 2 labels, both of which have a style set using this technique, but *different* styles (e.g., label 1 is red, label 2 is blue), are we going to see any overlap in the application of styles? What is the scope of a `Gtk.CssProvider`?

Vipul-Cariappa replied:
> If I'm reading this right, GTK styling is based on the idea that every app has a CSS stylesheet (possibly building on a "system theme" stylesheet); and then individual widgets can tweak that style by attaching an encoded style that defines a class, and attaching that class to the widget.

Yes. This is what is  happening.

> Would it make sense to use a name like toga as the custom class, rather than the generic class name?I'm thinking mostly for debugging purposes, being able to identify where a style change comes from. custom-style is sufficiently generic that it might not be obvious where the style is coming from.

I agree with you. This will be done in next commit.

> Is there any potential for namespace collision here? i.e., if I have 2 labels, both of which have a style set using this technique, but different styles (e.g., label 1 is red, label 2 is blue), are we going to see any overlap in the application of styles? What is the scope of a Gtk.CssProvider?

It two labels; label1 and label2 have different style sheets applied to them it will behave as expected. They will have different styles. 
If a css class custom-style is added to a label widget which sets the color to red and size to 20. Later if you add another class with the same name i.e custom-style which sets the color to black. Only the color will change and the size will remain 20.

freakboy3742 commented on src/gtk/toga_gtk/libs/style.py:
```diff
@@ -0,0 +1,49 @@
+    if value.size != -1:
+        style += f"font-size: {value.size}px; "
+
+    style += "}"
```
Generally, it's better to combine strings using `'\n'.join(parts)` or ''.join(parts), rather than a series of `+=` calls. Repeated addition requires a copy of the original string each time an addition is made, whereas joining a series of parts means a single copy is made.

Vipul-Cariappa replied:
If you look at `get_color_css` function and `get_bg_color_css` function I am not concatenating the strings. Python creates a single string at runtime. But in case of `get_font_css` function the default value for size is -1 due to which Gtk throws size can not be negative error. Therefore I am checking if size is -1, if so not adding them to the final css. Only in this specified use case I am concatenating the strings.

Vipul-Cariappa replied:
I understand your concern. Will change in next commit.

freakboy3742 commented on src/gtk/toga_gtk/widgets/button.py:
```diff
@@ -19,8 +25,23 @@ def set_label(self, label):
+        if value:
+            style_context = self.native.get_style_context()
+            css = get_font_css(value)
+            apply(style_context, css)
```
Is there any risk to adding these methods to the base widget, rather than explicitly redefining on each widget where we want to support color/font?

Vipul-Cariappa replied:
There is no risk in adding these methods to base widget from my testing. But it does not make sense for widgets like canvas and divider to have font or color. And even in winform back end the implementation is added in each widget class and not in base widget class.
But if you fell like it should be added at base class I have no problems with it.

freakboy3742 replied:
The difference with Winforms is that the Winforms backend requires an explicit "set_background_color()"-style API call, which literally does not exist for canvas/divider etc. As a result, having a "default" implementation will raise a hard error if it is used on a widget that doesn't have the necessary API. 

In the case of GTK, we're setting a style directive; if the widget doesn't support the style, it will be ignored (and, hypothetically, if GTK Canvas ever *did* start to support setting a background color, then it would immediately pick up the style Toga provides). 

That said - thinking about it more, `set_color` et al may need to exist to satisfy the API contract of the backend, in which case moving the implementation to the base class may be a non-starter. My key concern is that, as stands, there's a non-trivial, and essentially identical implementation duplicated across three separate subclasses (and potentially more in future). #1195 describes a general cleanup we'd like to do in this space across all backends; if we're adding new functionality, it makes sense to do this cleanup at the time the feature is added.

Vipul-Cariappa replied:
I will shift the `set_color`, `set_font` & `set_background_color` to base widget class in the next commit.
I would like to point out that you can not set background color to` Gtk.Fixed` (`toga.Box`) widget by this method. According to the research I did about it on the internet, looks like `Gtk.Fixed` does not support changing background color.

freakboy3742 reviewed (changes requested):
Looks good - the code organization is much better in this pass. I've flagged a couple of issues inline; I think they're either minor, or requests for clarification/confirmation.

I've also pushed an update to your branch that merges with the current master. Without that update, the PR review included commits that were related to other people's work (e.g., the deprecation of Python 3.5).

Vipul-Cariappa commented on src/gtk/toga_gtk/libs/style.py:
```diff
@@ -0,0 +1,49 @@
+from . import Gtk
+
+
+def apply(style_context, style):
```
Will be doing it in next commit.

Vipul-Cariappa commented:
@freakboy3742 I don't know a proper way to fix the error caused by the test suit. As I have shifted the `set_font` method to base class changing font style in `TextInput` widget is possible therefore I removed the `set_font` method from `TextInput` class. I need help to solve this [problem](https://github.com/beeware/toga/pull/1203/checks?check_run_id=2002236827#step:5:1619).

freakboy3742 commented:
There are 2 test failures:
* GTK's TextInput widget isn't satisfying the API contract for a TextInput widget - it's missing a `set_font()` call. The API contract is defined by the Dummy backend; every backend needs to adhere to that contract. This is what I warned about in my last code review - if the API contract specifies that a method needs to exist on a widget, it's difficult to abstract into the base class.
* GTK's base Widget has a set_font() method, but that method *doesn't* have an argument `font`. This is because you've renamed the argument `value`.

Vipul-Cariappa commented:
@freakboy3742 Thank You for the details. I have change the required. Hope this is helpful.

Vipul-Cariappa commented:
@freakboy3742 I guess the PR is ready to be merged.

federicotorrielli commented:
@freakboy3742 Do we have any plans to merge this asap? Not being able to modify colors on widget is really sad for Toga users.
